4단계 - 세션 구현하기

willing1026 opened the pull request:
- 세션을 이용한 로그인 판단
   - 로그인 : attribute에 User 객체 등록
   - 로그아웃 : attribute invalidate로 user객체 삭제

- 기타 코드 리팩토링

willing1026 committed:
# add : HttpSession 객체, Session관리 객체  추가

### add
- HttpSession : session객체
- HttpSessionManager : session 등록, 보관 객체

willing1026 committed:
# add & modify : HttpSession을 이용한 로그인 기능 구현, 코드 리팩토링

### HttpSession 이용한 로그인 기능 구현
- UserLoginController : 로그인 성공할 경우 세션ID 생성, user객체를 attribute에 등록
- UserListController : 로그인여부 확인방법 변경 ( cookie -> session )
- HttpSessionManager : login기능구현 (sessionId생성, attribute 추가),  세션삭제 기능 추가

### 기능 리팩토링
- HttpStringUtils : method 추가 (sessionId 추출, 특정값 추출, concate), method명 수정 ( splitAndFindByIndex -> substring )
- RequestMapping : controller 객체 재사용
- HttpRequestHeader : requestHeader parsing 부분 method로 추출
- HttpResponseResolver : method 분리 ( cookie가 있는경우, 없는경우 )

### 인스턴스명 정리
- HttpStringType : 인스턴스명 정리, 추가
- HttpRequestLine : enum인자명 변경에 따른 수정
- HttpRequestParams : enum인자명 변경에 따른 수정

- RequestHeader : 인스턴스명 변경 logger -> log
- ResponseHandler : 인스턴스명 변경 logger -> log
- WebServer : 인스턴스명 변경 logger -> log

### 불필요한 인자 제거
- ForwardController : 불필요한 import 제거
- HomeController : 불필요한 import 제거
- UserCreateController :  불필요한 import 제거, redirect 호출 부분 개선 (불필요한 null인자 제거)

### String Concatenate 방식 변경
- ResourceMapping : string concatenate 방식 변경
- HttpResponse : string concatenate 방식 변경
- HttpSession : string concatenate 방식 변경

### 테스트를 위한 null인자 converter 추가
- NullableConverter : csvResource의 null 값 인식을 위해 converter 추가

willing1026 committed:
# add & refactor : session을 이용한 logout기능 구현

### logout 기능
- UserLogoutController : session의 attribute 초기화 후 index.html로 redirect
- HttpSessionManager : logout method 기능 추가
- RequestMapping : logout controller 정보 추가
- index.html, form.html, list.html, profile.html : logout 기능 활성화

### session 추출 기능 refactoring
- HttpSessionManager : sessionId를 받아 처리하도록 변경 (requestHeader 받는 method 제거)
- HttpRequest : sessionId 구하는 method 추가
- HttpRequestHeader : null 예외처리 제외 ( method 호출 부분에서 관리하도록 변경 )
- UserListController : 관련 기능 refactoring

### 기타 refactoring
- HttpRequestLine : httpMethod enum 사용하도록 개선

javajigi commented on src/main/java/webserver/http/session/HttpSessionManager.java:
```diff
@@ -0,0 +1,57 @@
+        return sessions.get(httpSession.getId());
+    }
+
+    public static HttpSession login(Object attribute) {
```
HttpSessionManager는 전체 HttpSession을 관리하는 객체인데 애플리케이션 로직이라 할 수 있는 login과 logout 메소드를 가지는 것은 단일 책임원칙에 어긋 나는 것이 아닐까?

애플리케이션 로직과 Http 처리 로직을 분리할 것을 추천

willing1026 replied:
HttpSessionManager에서 애플리케이션 로직 제거 했습니다.
애플리케이션 로직은 Controller에 구현했습니다.

javajigi commented on src/test/java/webserver/http/session/HttpSessionTest.java:
```diff
@@ -0,0 +1,72 @@
+        String value = "rainy";
+        httpSession.setAttribute(name, value);
+
+        assertThat(httpSession.invalidate()).isTrue();
```
invalidate() 후에 해당 데이터가 존재하지 않는지도 체크해야하지 않을까?

willing1026 replied:
추가했습니다!

javajigi commented on src/test/java/webserver/http/session/HttpSessionTest.java:
```diff
@@ -0,0 +1,72 @@
+
+        String name = "weather";
+        String value = "rainy";
+        httpSession.setAttribute(name, value);
```
name, value를 재사용하지 않는데 굳이 로컬 변수로 구현할 필요가 있을까?

willing1026 replied:
아.. 해당 케이스 수정했습니다.

javajigi commented on src/test/java/webserver/http/session/HttpSessionTest.java:
```diff
@@ -0,0 +1,72 @@
+        User user = new User("id", "pw", "jw", "jw@gmail.com");
+        httpSession.setAttribute(user.getUserId(), user);
+
+        assertThat(httpSession.removeAttribute(user.getUserId())).isTrue();
```
attribute 삭제 후에 해당 데이터가 잘 삭제됐는지 확인하는 assert 문이 필요하지 않을까?

willing1026 replied:
네 추가했습니다!

javajigi reviewed (changes requested):
세션 구현은 잘 했네요. 💯 
단, SessionManager의 역할과 테스트 측면에서 피드백 좀 남겼어요.

willing1026 committed:
#modify : HttpSessionManager 애플리케이션로직 제거. HttpSession 테스트 케이스 수정

- HttpSessionManager : 애플리케이션로직(login, logout) 제거. Controller에 구현
- UserLoginController : 세션 관련 기능 구현
- UserLogoutController : 세션 관련 기능 구현

그 외 테스트케이스 개선, 불필요한 케이스 제거

javajigi reviewed (approved):
피드백 반영하느라 수고했어요. 💯
